src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -227,25 +274,138 @@ private static String extractQualifiedName(DetailAST ast) {
     }
 
     /**
-     * Register to outer super classes of given classAst that
+     * Register to outer super class of given classAst that
      * given classAst is extending them.
      *
-     * @param classAst class which outer super classes will be
+     * @param classAst class which outer super class will be
      *                 informed about nesting subclass
+     * @param qualifiedClassName qualifies class name(with package) of the current class
      */
-    private void registerNestedSubclassToOuterSuperClasses(DetailAST classAst) {
+    private void registerNestedSubclassToOuterSuperClasses(DetailAST classAst,
+                                                           String qualifiedClassName) {
         final String currentAstSuperClassName = getSuperClassName(classAst);
+
         if (currentAstSuperClassName != null) {
+            String classToBeRegisteredAsSuperClass = currentAstSuperClassName;
+
+            final List<ClassDesc> sameNamedClassList = classesWithSameName(
+                    currentAstSuperClassName);
+            if (!sameNamedClassList.isEmpty()) {
+                classToBeRegisteredAsSuperClass = getClassToBeRegisteredAsSuperClass(
+                        qualifiedClassName, sameNamedClassList);
+            }
+
             for (ClassDesc classDesc : classes) {
                 final String classDescQualifiedName = classDesc.getQualifiedName();
-                if (doesNameInExtendMatchSuperClassName(classDescQualifiedName,
-                        currentAstSuperClassName)) {
+                if (classToBeRegisteredAsSuperClass.equals(classDescQualifiedName)) {
                     classDesc.registerNestedSubclass();
+                    break;
                 }
             }
         }
     }
 
+    /**
+     * Checks if there is a class with same name.
+     *
+     * @param className name of the class
+     * @return true if there is another class with same name.
+     */
+    private List<ClassDesc> classesWithSameName(String className) {
+        return classes.stream()
+                .filter(classDesc -> {
+                    return classDesc.getQualifiedName().endsWith(
+                            PACKAGE_SEPARATOR + className);
+                })
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Get name of the class to be registered as super class.
+     * This method after getting the {@link ClassDesc#classNameMatchingCount}
+     * compares the classes on basis of it. It at a time compares two classes
+     * and keeps updating {@code classToBeRegisteredAsSuperClass} on basis of it.
+     *
+     * @param currentClassName current class name
+     * @param classesWithSameName classes which same name as super class
+     * @return name of the class to be registered as super class
+     */
+    private static String getClassToBeRegisteredAsSuperClass(String currentClassName,
```

Comment:
@Vyom-Yadav I have an idea. This method should ignore all classes that are more nested than the `currentClassName`. This code will not compile:
```
    class K extends F { // class  X.F is not accessible here
    }

    class X {
        class F {}
        }
    }
```

Reply:
So, the method may be rewritten like this:
```java
    private static ClassDesc getClassToBeRegisteredAsSuperClass(String currentClassName,
                                                             List<ClassDesc> classesWithSameName) {
        int maxMatchingCount = 0;
        ClassDesc classToBeRegisteredAsSuperClass = null;
        for (ClassDesc desc : classesWithSameName) {
            final String classWithSameName = desc.getQualifiedName();
            if (isFormerClassNestedMoreThanTheLatterClass(classWithSameName, currentClassName)) {
                continue;
            }
            int matchingCount = classNameMatchingCount(currentClassName, classWithSameName);
            if (matchingCount > maxMatchingCount) {
                maxMatchingCount = matchingCount;
                classToBeRegisteredAsSuperClass = desc;
            }
        }
        return classToBeRegisteredAsSuperClass;
}
```